abb_hardware_interface/src/rws_state_publisher_ros.cpp

```diff
@@ -0,0 +1,138 @@
+ *    * Redistributions in binary form must reproduce the
+ *      above copyright notice, this list of conditions
+ *      and the following disclaimer in the documentation
+ *      and/or other materials provided with the
+ *      distribution.
+ *    * Neither the name of ABB nor the names of its
+ *      contributors may be used to endorse or promote
+ *      products derived from this software without
+ *      specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ***********************************************************************************************************************
+ */
+
+// This file is a modified copy from
+// https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_rws_service_provider/src
+// https://github.com/ros-industrial/abb_robot_driver/tree/master/abb_rws_state_publisher/src
+
+#include "abb_hardware_interface/rws_state_publisher_ros.hpp"
+
+#include "abb_hardware_interface/mapping.hpp"
+#include "abb_hardware_interface/utilities.hpp"
+
+namespace {
+/**
+ * \brief Time [s] for throttled ROS logging.
+ */
+constexpr double THROTTLE_TIME{10.0};
+}  // namespace
+
+namespace abb_rws_client {
+RWSStatePublisherROS::RWSStatePublisherROS(const rclcpp::Node::SharedPtr& node, const std::string& robot_ip,
+                                           unsigned short robot_port)
+    : RWSClient(node, robot_ip, robot_port) {
+  node_->declare_parameter("polling_rate", 5.0);
+  abb::robot::initializeMotionData(motion_data_, robot_controller_description_);
+
+  auto sensor_qos =
+      rclcpp::QoS(rclcpp::QoSInitialization(rmw_qos_profile_sensor_data.history, rmw_qos_profile_sensor_data.depth),
+                  rmw_qos_profile_sensor_data);
+  joint_state_pub_ = node_->create_publisher<sensor_msgs::msg::JointState>("~/joint_states", sensor_qos);
+
+  system_state_pub_ = node_->create_publisher<abb_robot_msgs::msg::SystemState>("~/system_state", 1);
+
+  if (abb::robot::utilities::verify_state_machine_add_in_presence(robot_controller_description_.system_indicators())) {
+    runtime_state_pub_ =
+        node_->create_publisher<abb_rapid_sm_addin_msgs::msg::RuntimeState>("~/sm_addin/runtime_state", 1);
+  }
```

Comment:
Does this need to use the `timer_` variable you declared in the header, that uses `timer_callback` below? Otherwise, I don't see how `timer_callback` is used.

Reply:
I assume the `polling_rate` parameter was declared earlier for use with a timer as well...?